app/models/user.rb

```diff
@@ -1,109 +1,169 @@
-        @users.each do |friend|
-          Friendship.where(user: user_id, friend_id: friend.id).first_or_create
-        end
-      end
+	# Include default devise modules. Others available are:
+	# :confirmable, :lockable, :timeoutable
+	devise :database_authenticatable, :registerable, :recoverable, :rememberable, :trackable, :validatable, :omniauthable, omniauth_providers: [:facebook]
+
+	has_many :friendships
+	has_many :users, through: :friendships
+	has_many :reviews, dependent: :destroy
+	has_many :movies, dependent: :destroy
+	has_many :visits, class_name: "Ahoy::Visit"
+	validates_uniqueness_of :email
+	enum access_level: [:user, :admin, :super_admin]
+
+	def self.new_with_session(params, session)
+		super.tap do |user|
+			if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]
+				user.email = data["email"] if user.email.blank?
+			end
+		end
+	end
+
+	def self.from_omniauth(auth)
+
+		user = User.where(provider: auth.provider, uid: auth.uid).first
+		if user.blank?
+
+			# immediately get 60 day auth token
+			oauth = Koala::Facebook::OAuth.new(ENV["FACEBOOK_APP_ID"], ENV["FACEBOOK_SECRET"])
+			new_access_info = oauth.exchange_access_token_info auth.credentials.token
+
+			new_access_token = new_access_info["access_token"]
+			# Facebook updated expired attribute
+			new_access_expires_at = DateTime.now + new_access_info["expires_in"].to_i.seconds
+
+			user = User.new
+			user.provider = auth.provider
+			user.uid = auth.uid
+			user.email = auth.info.email
+			user.password = Devise.friendly_token[0,20]
+			user.name = auth.info.name   # assuming the user has a name
+			user.image = auth.info.image # assuming the user has an image
+			user.gender = auth.extra.raw_info.gender # assuming the user has a gender
+			user.access_level = 0
+			user.link = auth.extra.raw_info.link
+			unless auth.extra.raw_info.location.nil?
+				user.location = auth.extra.raw_info.location['name'] # assuming the user has a location
+			end
+			unless auth.extra.raw_info.hometown.nil?
+				user.hometown = auth.extra.raw_info.hometown['name'] # assuming the user has a hometown
+			end
+			user.token = new_access_token # originally auth.credentials.token
+			user.token_expires_at = new_access_expires_at # originally Time.at(auth.credentials.expires_at)
+			user.save(validate: false)
+		else
+			# check access_token
+			user.access_token
+		end
+		user
+	end
+
+	def add_movies
+		@facebook ||= client
+		@movies = @facebook.get_object('me/movies/', {}, api_version: 'v3.1')
+		puts @movies
+		loop do
+			break if @movies.blank?
+
+			help_add_movies
+			@movies = @movies.next_page
+		end
+	end
+
+	def add_friends
+		@facebook ||= client
+		@friends = @facebook.get_connections('me', 'friends')
+		puts @friends
+		loop do
+			break if @friends.blank?
+
+			help_add_friends
+			@friends = @friends.next_page
+		end
+	end
+
+	def help_add_movies
+		movie_names = @movies.collect { |f| f['name'] }
```

Comment:
You should be able to use `@movies.pluck(:name)` instead
https://apidock.com/rails/ActiveRecord/Calculations/pluck

Reply:
@jprokay this didn't work. I'm not sure why. When I do ` @movies = @graph.get_object('me?fields=movies', {}, api_version: 'v3.1')`, and print out `@movies` I get this > https://pastebin.com/XbpMgVT7. How would I get the name for each movie in the list? I have the same issue in `help_add_friends` too

Reply:
I have pushed an update so that help_add_movies works, but I still can't get your recommendation to work. https://github.com/sdsunjay/movie-recommendations-with-ml/blob/ce0ff07c706299d69cd7476dad353b67519e26a0/app/models/user.rb#L109